### ISSUE TITLE
simulate scope at let to a if statement but failed

### ISSUE THREAD

**Issue by arex0** [microsoft/TypeScript#36853]
**TypeScript Version:**  3.7.5

**Search Terms:**  if-let, scope, closure

**Expected behavior:** tsc report no error

**Actual behavior:** tsc report "Variable 'a' is used before being assigned."

**Related Issues:**

**Code**
if-let statement is easy to control scope in golang
```go
var a A 
if i := bytes.Index(s,sep); i != -1 {
  a = makeA(i) // use i here
}else{
  a = make(A)
  // others code do not use i here
}
a.doSomthing()
```
I try to use closure to simulate it, because ts can not use if-let statement. tsc reported error "Variable 'a' is used before being assigned."
```ts
class A {
    i: number | undefined
    constructor(i?: number) {
        if (i !== undefined) {
            this.i = i
        }
    }
    doSomething() {
        if (this.i === undefined) console.log("A init in else statement")
        else console.log("A init in if statement")
    }
}

function f(sep: number) {
    let a: A
    if (((i: number) => {
        if (i == -1) return false
        a = new A(i) // use i there
        return true
    })(arr.indexOf(sep))) {
    } else {
        a = new A
        // others code do not use i there
    }
    a.doSomething()
}

let arr = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

f(10)
f(5)
```

<details><summary><b>Output</b></summary>

```ts
"use strict";
class A {
    constructor(i) {
        if (i !== undefined) {
            this.i = i;
        }
    }
    doSomething() {
        if (this.i === undefined)
            console.log("A init in else statement");
        else
            console.log("A init in if statement");
    }
}
function f(sep) {
    let a;
    if (((i) => {
        if (i == -1)
            return false;
        a = new A(i); // use i there
        return true;
    })(arr.indexOf(sep))) {
    }
    else {
        a = new A;
        // others code do not use i there
    }
    a.doSomething();
}
let arr = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9];
f(10);
f(5);

```

</details>

<details><summary><b>Compiler Options</b></summary>

```json
{
  "compilerOptions": {
    "noImplicitAny": true,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "strictPropertyInitialization": true,
    "strictBindCallApply": true,
    "noImplicitThis": true,
    "noImplicitReturns": true,
    "useDefineForClassFields": false,
    "alwaysStrict": true,
    "allowUnreachableCode": false,
    "allowUnusedLabels": false,
    "downlevelIteration": false,
    "noEmitHelpers": false,
    "noLib": false,
    "noStrictGenericChecks": false,
    "noUnusedLocals": false,
    "noUnusedParameters": false,
    "esModuleInterop": true,
    "preserveConstEnums": false,
    "removeComments": false,
    "skipLibCheck": false,
    "checkJs": false,
    "allowJs": false,
    "declaration": true,
    "experimentalDecorators": false,
    "emitDecoratorMetadata": false,
    "target": "ES2017",
    "module": "ESNext"
  }
}
```

</details>

**Playground Link:** [Provided](https://www.typescriptlang.org/play/?noLib=true&noStrictGenericChecks=true&noUnusedLocals=true&noUnusedParameters=true&experimentalDecorators=true&emitDecoratorMetadata=true#)

**Comment by jcalz** [comment]
The new issue template asks for a Playground link https://www.typescriptlang.org/play/ where your issue is reproduced.  The above code has [several errors](https://www.typescriptlang.org/play/#code/DYUwLgBAhgXBCCAoAlgMwgCi8mA7ArgLYBGIATgJQC8AfAN6IQRqbIRVUQC0AjBRGXD4yuCKijAAziEbR2EXCADuCDMn4B6DRHzTmEMAAtyMpoLDDRYMvhkBfChmIBPMCEkA6AJK4AJiAAPDEkAGmkABwoohjsQKRAGJih5RRUkJi0IAHsjckkIAGMs-whfLIUcnT02XMFEO0QoDzKAZSzCcENkXABzDH4gA) due to missing declarations, but not the one you're reporting.   Please consider cleaning it up if you don't want this closed as unactionable.  Good luck!

**Comment by arex0** [comment]
@jcalz thank you for you commented, it is clean up now

**Comment by MartinJohns** [comment]
Duplicate of #9998.

**Comment by jcalz** [comment]
I don't know about #9998; the function here is an IIFE which, IIRC (ha ha), is inlined:

```ts
let z: string;
(() => { z = "a" })()
z.toUpperCase(); // okay
```

Instead I think it has to do with trying to "save" the results of CFA in variables, return them from functions, etc., and then use them later... which doesn't work, as in #12184, #20497, #24865 (and probably others):

```ts
let a: string;
const bool = Math.random() < 0.5;
if (bool) { a = ""; }
if (!bool) { a = ""; }
a.toUpperCase(); // error
```

In any case, a mcve for this probably looks like:

```ts
let b: string;
if ((() => {
    if (Math.random() < 0.5) return true;
    b = "";
    return false;
})()) {
    b = "";
}
b.toUpperCase(); // error
```


[Playground link](https://www.typescriptlang.org/play/#code/DYUwLgBAXgXBDOYBOBLAdgcwNwCgAUeAlBALwB8EA3tKRAEQCGdEAvoUTlAHRgD2AqgAdBIJAGEG8EESwQA9HIi8A1gwCeOCDlCQGcRKky4AxrzSIIAI169gtALIMwACy5IGaACa8AtkQgAPBAADFwArLgoAGYQeNa2xNQMtHR0siw40bEAhPHAiRDJJPRprDgMPALCohJSMvKKoki8SDja4Fb6yOjYmTEE-uRUmhCjWXiOLm4e3n7EQaFhxEjgAK5IaBDIqyC4o6OWKWkjoytg65tRDMBSuGxEiSdWR3c4lpVCIuKS0oSyChAmi02kA)

**Comment by RyanCavanaugh** [comment]
We don't have any kind of mechanisms capable of creating inferences like "This function either returns `true` *and* makes an assignment, *or* returns `false` when it doesn't", nor frequently-encountered use cases in which case this kind of reasoning is necessary.
